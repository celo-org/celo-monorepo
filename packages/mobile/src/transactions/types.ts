import { TokenTransactionType } from 'src/apollo/types'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { v4 as uuidv4 } from 'uuid'

export interface ExchangeStandby {
  context: TransactionContext
  type: TokenTransactionType.Exchange
  status: TransactionStatus
  inSymbol: CURRENCY_ENUM
  inValue: string
  outSymbol: CURRENCY_ENUM
  outValue: string
  timestamp: number
  hash?: string
}

export interface TransferStandby {
  context: TransactionContext
  type: TransferTransactionType
  status: TransactionStatus
  value: string
  comment: string
  symbol: CURRENCY_ENUM
  timestamp: number
  address: string
  hash?: string
}

export type StandbyTransaction = ExchangeStandby | TransferStandby

// Context used for logging the transaction execution flow.
export interface TransactionContext {
  // Unique identifier used for tracking a transaction within logging.
  // Note that this is not the transaction hash, which is unknown when creating a new transaction.
  id: string

  // A tag provided by the caller to provide context on the purpose.
  tag?: string

  // A short contextual description of what the transaction does. (e.g. "Approve attestations")
  description?: string
}

export function newTransactionContext(tag: string, description: string) {
  return {
    id: uuidv4(),
    tag,
    description,
  }
}

export enum TransactionStatus {
  Pending = 'Pending',
  Complete = 'Complete',
  Failed = 'Failed',
}

type TransferTransactionType =
  | TokenTransactionType.Sent
  | TokenTransactionType.Received
  | TokenTransactionType.EscrowReceived
  | TokenTransactionType.EscrowSent
  | TokenTransactionType.Faucet
  | TokenTransactionType.VerificationReward
  | TokenTransactionType.VerificationFee
  | TokenTransactionType.InviteSent
  | TokenTransactionType.InviteReceived
  | TokenTransactionType.NetworkFee
