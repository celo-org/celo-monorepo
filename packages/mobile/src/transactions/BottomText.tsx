import colors from '@celo/react-components/styles/colors.v2'
import fontStyles from '@celo/react-components/styles/fonts.v2'
import React from 'react'
import { StyleSheet, Text } from 'react-native'

interface BottomTextProps {
  children: React.ReactNode
}

export default function BottomText({ children }: BottomTextProps) {
  return <Text style={styles.text}>{children}</Text>
}

const styles = StyleSheet.create({
  text: {
    ...fontStyles.regular,
    color: colors.gray4,
    textAlign: 'center',
    marginTop: 'auto',
  },
})
