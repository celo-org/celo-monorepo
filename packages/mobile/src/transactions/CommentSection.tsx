import HorizontalLine from '@celo/react-components/components/HorizontalLine'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts.v2'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import { Namespaces } from 'src/i18n'

interface Props {
  comment?: string | null
}

export default function CommentSection({ comment }: Props) {
  const { t } = useTranslation(Namespaces.sendFlow7)
  if (!comment) {
    return null
  }

  return (
    // Uses View instead of Fragment to workaround a glitch with LayoutAnimation
    <View>
      <HorizontalLine />
      <Text style={styles.sectionLabel}>{t('commentLabel')}</Text>
      <Text style={styles.comment}>{comment}</Text>
    </View>
  )
}

const styles = StyleSheet.create({
  sectionLabel: {
    ...fontStyles.label,
    color: colors.gray3,
    marginBottom: 4,
  },
  comment: {
    ...fontStyles.regular,
  },
})
