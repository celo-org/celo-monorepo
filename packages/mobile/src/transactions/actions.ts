import { ExchangeConfirmationCardProps } from 'src/exchange/ExchangeConfirmationCard'
import i18n from 'src/i18n'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { ConfirmationInput as SendConfirmationCardProps } from 'src/send/SendConfirmation'
import { TransferConfirmationCardProps } from 'src/send/TransferConfirmationCard'
import { StandbyTransaction, TransactionTypes } from 'src/transactions/reducer'
import { web3 } from 'src/web3/contracts'

export enum Actions {
  ADD_STANDBY_TRANSACTION = 'TRANSACTIONS/ADD_STANDBY_TRANSACTION',
  REMOVE_STANDBY_TRANSACTION = 'TRANSACTIONS/REMOVE_STANDBY_TRANSACTION',
  RESET_STANDBY_TRANSACTIONS = 'TRANSACTIONS/RESET_STANDBY_TRANSACTIONS',
  ADD_HASH_TO_STANDBY_TRANSACTIONS = 'TRANSACTIONS/ADD_HASH_TO_STANDBY_TRANSACTIONS',
  TRANSACTION_CONFIRMED = 'TRANSACTIONS/TRANSACTION_CONFIRMED',
}

export interface AddStandbyTransaction {
  type: Actions.ADD_STANDBY_TRANSACTION
  transaction: StandbyTransaction
}

export interface RemoveStandbyTransaction {
  type: Actions.REMOVE_STANDBY_TRANSACTION
  idx: string
}

export interface ResetStandbyTransactions {
  type: Actions.RESET_STANDBY_TRANSACTIONS
}

export interface AddHashToStandbyTransaction {
  type: Actions.ADD_HASH_TO_STANDBY_TRANSACTIONS
  idx: string
  hash: string
}

export interface TransactionConfirmed {
  type: Actions.TRANSACTION_CONFIRMED
  txId: string
}

export type ActionTypes =
  | AddStandbyTransaction
  | RemoveStandbyTransaction
  | ResetStandbyTransactions
  | AddHashToStandbyTransaction

export const generateStandbyTransactionId = (recipientAddress: string) => {
  return web3.utils.sha3(recipientAddress + String(Date.now()))
}

export const addStandbyTransaction = (transaction: StandbyTransaction): AddStandbyTransaction => ({
  type: Actions.ADD_STANDBY_TRANSACTION,
  transaction,
})

export const removeStandbyTransaction = (idx: string): RemoveStandbyTransaction => ({
  type: Actions.REMOVE_STANDBY_TRANSACTION,
  idx,
})

export const resetStandbyTransactions = (): ResetStandbyTransactions => ({
  type: Actions.RESET_STANDBY_TRANSACTIONS,
})

export const transactionConfirmed = (txId: string): TransactionConfirmed => ({
  type: Actions.TRANSACTION_CONFIRMED,
  txId,
})

export const addHashToStandbyTransaction = (
  idx: string,
  hash: string
): AddHashToStandbyTransaction => ({
  type: Actions.ADD_HASH_TO_STANDBY_TRANSACTIONS,
  idx,
  hash,
})

export const navigateToPaymentTransferReview = (
  type: string,
  timestamp: number,
  confirmationProps: TransferConfirmationCardProps
) => {
  let headerText = ''
  switch (type) {
    case TransactionTypes.SENT:
      headerText = i18n.t('sendFlow7:sentPayment')
      break
    case TransactionTypes.RECEIVED:
      headerText = i18n.t('receiveFlow8:receivedPayment')
      break
    case TransactionTypes.VERIFICATION_FEE:
      headerText = i18n.t('walletFlow5:verificationFee')
      break
    case TransactionTypes.FAUCET:
      headerText = i18n.t('receiveFlow8:receivedDollars')
      break
    case TransactionTypes.INVITE_SENT:
      headerText = i18n.t('inviteFlow11:inviteComplete')
      break
    case TransactionTypes.INVITE_RECEIVED:
      headerText = i18n.t('inviteFlow11:inviteReceived')
      break
    case TransactionTypes.NETWORK_FEE:
      headerText = i18n.t('walletFlow5:networkFee')
      break
  }

  navigate(Screens.TransactionReview, {
    reviewProps: {
      type,
      timestamp,
      header: headerText,
    },
    confirmationProps,
  })
}

export const navigateToExchangeReview = (
  timestamp: number,
  confirmationProps: ExchangeConfirmationCardProps
) => {
  navigate(Screens.TransactionReview, {
    reviewProps: {
      type: TransactionTypes.EXCHANGE,
      timestamp,
      header: i18n.t('exchangeFlow9:exchange'),
    },
    confirmationProps,
  })
}

export const navigateToRequestedPaymentReview = (confirmationInput: SendConfirmationCardProps) => {
  navigate(Screens.SendConfirmation, { confirmationInput })
}
