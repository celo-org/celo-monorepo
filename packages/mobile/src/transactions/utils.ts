import i18n from 'src/i18n'
import { FeedItem } from 'src/transactions/TransactionFeed'
import { formatFeedSectionTitle, timeDeltaInDays } from 'src/utils/time'

// Groupings:
// Recent -> Last 7 days.
// [Current month] - "July" -> Captures transactions from the current month that aren’t captured in Recent.
// [Previous months] - "June" -> Captures transactions by month.
// [Months over a year ago] — "July 2019" -> Same as above, but with year appended.
// Sections are hidden if they have no items.
export const groupFeedItemsInSections = (feedItems: FeedItem[]) => {
  const sectionsMap: {
    [key: string]: {
      data: FeedItem[]
      daysSinceTransaction: number
    }
  } = {}

  feedItems.reduce((sections, item) => {
    const daysSinceTransaction = timeDeltaInDays(Date.now(), item.timestamp)
    const key =
      daysSinceTransaction <= 7
        ? i18n.t('walletFlow5:feedSectionHeaderRecent')
        : formatFeedSectionTitle(item.timestamp, i18n)
    sections[key] = sections[key] || {
      daysSinceTransaction,
      data: [],
    }
    sections[key].data.push(item)
    return sections
  }, sectionsMap)

  return Object.entries(sectionsMap)
    .sort((a, b) => a[1].daysSinceTransaction - b[1].daysSinceTransaction)
    .map(([key, value]) => ({
      title: key,
      data: value.data,
    }))
}
