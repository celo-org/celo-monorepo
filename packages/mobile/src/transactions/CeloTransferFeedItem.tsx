// VIEW that's used on the CELO screen activity feed for CELO transfers (in or out).

import Touchable from '@celo/react-components/components/Touchable'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import { CeloExchangeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { TransferItemFragment } from 'src/apollo/types'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import { formatShortenedAddress } from 'src/components/ShortenedAddress'
import { Namespaces } from 'src/i18n'
import { navigateToPaymentTransferReview } from 'src/transactions/actions'
import { TransactionStatus } from 'src/transactions/types'
import { getDatetimeDisplayString } from 'src/utils/time'

type Props = TransferItemFragment & {
  status?: TransactionStatus
}

export function CeloTransferFeedItem(props: Props) {
  const { t, i18n } = useTranslation(Namespaces.walletFlow5)
  const { address, amount, comment, status, timestamp, type } = props
  const onPress = () => {
    ValoraAnalytics.track(CeloExchangeEvents.celo_transaction_select)

    navigateToPaymentTransferReview(type, timestamp, {
      address,
      comment,
      amount,
      type,
      // fee TODO: add fee here.
    })
  }

  const dateTimeFormatted = getDatetimeDisplayString(timestamp, i18n)
  const isPending = status === TransactionStatus.Pending
  const isWithdrawal = new BigNumber(amount.value).isNegative()

  return (
    <Touchable onPress={onPress}>
      <View style={styles.container}>
        <View style={styles.firstRow}>
          <View style={styles.desc}>
            <Text style={styles.txMode}>
              {t(isWithdrawal ? 'feedItemGoldWithdrawal' : 'feedItemGoldReceived', {
                address: formatShortenedAddress(address),
              })}
            </Text>
          </View>
          <View>
            <CurrencyDisplay amount={amount} style={styles.amount} />
          </View>
        </View>
        <View style={styles.secondRow}>
          <Text style={styles.time}>{isPending ? t('confirmingExchange') : dateTimeFormatted}</Text>
        </View>
      </View>
    </Touchable>
  )
}

const styles = StyleSheet.create({
  container: {
    justifyContent: 'space-between',
    flex: 1,
    padding: variables.contentPadding,
  },
  firstRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    paddingBottom: 2,
  },
  desc: {
    flexDirection: 'row',
  },
  txMode: {
    ...fontStyles.regular500,
    color: colors.dark,
  },
  amount: {
    ...fontStyles.regular500,
    color: colors.dark,
  },
  time: {
    ...fontStyles.small,
    color: colors.gray4,
  },
  secondRow: {},
})

export default CeloTransferFeedItem
