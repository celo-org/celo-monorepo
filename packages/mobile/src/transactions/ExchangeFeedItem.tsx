import Touchable from '@celo/react-components/components/Touchable'
import ExchangeArrow from '@celo/react-components/icons/ExchangeArrow'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { Image, StyleSheet, Text, View } from 'react-native'
import { HomeExchangeFragment } from 'src/apollo/types'
import { CURRENCY_ENUM, resolveCurrency } from 'src/geth/consts'
import { Namespaces } from 'src/i18n'
import { navigateToExchangeReview } from 'src/transactions/actions'
import { ExchangeStandby, TransactionStatus } from 'src/transactions/reducer'
import { getMoneyDisplayValue } from 'src/utils/formatting'
import { formatFeedTime, getDatetimeDisplayString } from 'src/utils/time'

type Props = (HomeExchangeFragment | ExchangeStandby) &
  WithNamespaces & {
    showImage: boolean
    status?: TransactionStatus
  }

export function ExchangeFeedItem(props: Props) {
  const { showImage, t, inSymbol, inValue, status, outValue, outSymbol, timestamp, i18n } = props

  const onPress = () => {
    navigateToExchangeReview(timestamp, {
      makerToken: resolveCurrency(inSymbol),
      makerAmount: new BigNumber(inValue),
      takerAmount: new BigNumber(outValue),
    })
  }

  const inCurrency = resolveCurrency(inSymbol)
  const outCurrency = resolveCurrency(outSymbol)
  const dollarAmount = inCurrency === CURRENCY_ENUM.DOLLAR ? inValue : outValue
  const dollarDirection = inCurrency === CURRENCY_ENUM.DOLLAR ? '-' : ''
  const timeFormatted = formatFeedTime(timestamp, i18n)
  const dateTimeFormatted = getDatetimeDisplayString(timestamp, t, i18n)
  const isPending = status === TransactionStatus.Pending

  const inStyle = {
    color: isPending
      ? colors.gray
      : inCurrency === CURRENCY_ENUM.DOLLAR
        ? colors.celoGreen
        : colors.celoGold,
  }

  const outStyle = {
    color: isPending
      ? colors.gray
      : outCurrency === CURRENCY_ENUM.DOLLAR
        ? colors.celoGreen
        : colors.celoGold,
  }

  return (
    <Touchable onPress={onPress}>
      <View style={styles.container}>
        {showImage && (
          <View style={styles.imageContainer}>
            <Image
              source={
                inCurrency === CURRENCY_ENUM.DOLLAR
                  ? require(`src/transactions/ExchangeGreenGold.png`)
                  : require(`src/transactions/ExchangeGoldGreen.png`)
              }
              style={styles.image}
              resizeMode="contain"
            />
          </View>
        )}
        <View style={[styles.contentContainer, showImage && styles.contentContainerWithImage]}>
          <View style={styles.titleContainer}>
            <Text style={fontStyles.bodySmallSemiBold}>{t('exchange')}</Text>
            <Text
              style={[
                dollarDirection === '-'
                  ? fontStyles.activityCurrencySent
                  : fontStyles.activityCurrencyReceived,
                styles.amount,
              ]}
            >
              {dollarDirection}
              {getMoneyDisplayValue(dollarAmount)}
            </Text>
          </View>
          <View style={styles.exchangeContainer}>
            <Text style={[fontStyles.activityCurrency, inStyle]}>
              {getMoneyDisplayValue(inValue)}
            </Text>
            <View style={styles.arrow}>
              <ExchangeArrow />
            </View>
            <Text style={[fontStyles.activityCurrency, outStyle]}>
              {getMoneyDisplayValue(outValue)}
            </Text>
          </View>
          <View style={styles.statusContainer}>
            {isPending && (
              <Text style={[fontStyles.bodySmall, styles.transactionStatus]}>
                <Text style={[fontStyles.bodySmallBold, styles.textPending]}>
                  {t('confirmingExchange')}
                </Text>
                {' ' + timeFormatted}
              </Text>
            )}
            {status === TransactionStatus.Complete && (
              <Text style={[fontStyles.bodySmall, styles.transactionStatus]}>
                {dateTimeFormatted}
              </Text>
            )}
            {status === TransactionStatus.Failed && (
              <Text style={[fontStyles.bodySmall, styles.transactionStatus]}>
                <Text style={fontStyles.linkSmall}>{t('exchangeFailed')}</Text>
                {' ' + timeFormatted}
              </Text>
            )}
          </View>
        </View>
      </View>
    </Touchable>
  )
}

const styles = StyleSheet.create({
  arrow: {
    paddingHorizontal: 5,
    paddingBottom: 8,
  },
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    padding: variables.contentPadding,
  },
  imageContainer: {
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    height: 40,
    width: 40,
  },
  image: {
    width: 40,
  },
  contentContainer: {
    flex: 1,
  },
  contentContainerWithImage: {
    marginLeft: variables.contentPadding,
  },
  titleContainer: {
    flexDirection: 'row',
    marginTop: 3,
  },
  amount: {
    marginLeft: 'auto',
    paddingLeft: 10,
  },
  exchangeContainer: {
    // alignSelf: 'flex-start',
    alignItems: 'center',
    flexDirection: 'row',
    // justifyContent: 'flex-end',
  },
  statusContainer: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: 8,
  },
  textPending: {
    fontSize: 13,
    lineHeight: 18,
    color: colors.celoGreen,
  },
  transactionStatus: {
    color: '#BDBDBD',
  },
  localAmount: {
    marginLeft: 'auto',
    paddingLeft: 10,
    fontSize: 14,
    lineHeight: 18,
    color: '#BDBDBD',
  },
})

export default withNamespaces(Namespaces.walletFlow5)(React.memo(ExchangeFeedItem))
