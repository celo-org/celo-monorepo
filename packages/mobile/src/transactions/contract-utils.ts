import { values } from 'lodash'
import { estimateGas } from 'src/web3/utils'
import { Tx } from 'web3-core'
import { TransactionObject, TransactionReceipt } from 'web3-eth'

export type TxLogger = (event: SendTransactionLogEvent) => void

export function emptyTxLogger(_event: SendTransactionLogEvent) {
  return
}

interface TxPromiseResolvers {
  receipt: (receipt: TransactionReceipt) => void
  transactionHash: (transactionHash: string) => void
  confirmation: (confirmation: boolean) => void
}

type PromiseRejection = (error: Error) => void
interface TxPromiseReject {
  receipt: PromiseRejection
  transactionHash: PromiseRejection
  confirmation: PromiseRejection
}

export interface TxPromises {
  receipt: Promise<TransactionReceipt>
  transactionHash: Promise<string>
  confirmation: Promise<boolean>
}

export function awaitConfirmation(txPromises: TxPromises) {
  return txPromises.confirmation
}

// Couldn't figure out how to make it generic
export type SendTransaction<T> = (
  tx: TransactionObject<any>,
  account: string,
  txId?: string
) => Promise<T>

export enum SendTransactionLogEventType {
  Started,
  EstimatedGas,
  ReceiptReceived,
  TransactionHashReceived,
  Confirmed,
  Failed,
  Exception,
}

interface Started {
  type: SendTransactionLogEventType.Started
}
const Started: Started = { type: SendTransactionLogEventType.Started }

interface Confirmed {
  type: SendTransactionLogEventType.Confirmed
}
const Confirmed: Confirmed = { type: SendTransactionLogEventType.Confirmed }

export type SendTransactionLogEvent =
  | Started
  | EstimatedGas
  | ReceiptReceived
  | TransactionHashReceived
  | Confirmed
  | Failed
  | Exception

interface EstimatedGas {
  type: SendTransactionLogEventType.EstimatedGas
  gas: number
}

function EstimatedGas(gas: number): EstimatedGas {
  return { type: SendTransactionLogEventType.EstimatedGas, gas }
}

interface ReceiptReceived {
  type: SendTransactionLogEventType.ReceiptReceived
  receipt: TransactionReceipt
}

function ReceiptReceived(receipt: TransactionReceipt): ReceiptReceived {
  return { type: SendTransactionLogEventType.ReceiptReceived, receipt }
}

interface TransactionHashReceived {
  type: SendTransactionLogEventType.TransactionHashReceived
  hash: string
}

function TransactionHashReceived(hash: string): TransactionHashReceived {
  return { type: SendTransactionLogEventType.TransactionHashReceived, hash }
}

interface Failed {
  type: SendTransactionLogEventType.Failed
  error: Error
}

function Failed(error: Error): Failed {
  return { type: SendTransactionLogEventType.Failed, error }
}

interface Exception {
  type: SendTransactionLogEventType.Exception
  error: Error
}

function Exception(error: Error): Exception {
  return { type: SendTransactionLogEventType.Exception, error }
}

/**
 * sendTransactionAsync mainly abstracts the sending of a transaction in a promise like
 * interface. Use the higher-order sendTransactionFactory as a consumer to configure
 * logging and promise resolution
 * TODO: Should probably renamed to sendTransaction once we remove the current
 *       sendTransaction
 * @param tx The transaction object itself
 * @param account The address from which the transaction should be sent
 * @param feeCurrencyContract The contract instance of the Token in which to pay gas for
 * @param logger An object whose log level functions can be passed a function to pass
 *               a transaction ID
 */
export async function sendTransactionAsync<T>(
  tx: TransactionObject<T>,
  account: string,
  feeCurrencyAddress: string,
  logger: TxLogger = emptyTxLogger,
  estimatedGas?: number,
  gasPrice?: string,
  nonce?: number
): Promise<TxPromises> {
  // @ts-ignore
  const resolvers: TxPromiseResolvers = {}
  // @ts-ignore
  const rejectors: TxPromiseReject = {}

  const receipt: Promise<TransactionReceipt> = new Promise((resolve, reject) => {
    resolvers.receipt = resolve
    rejectors.receipt = reject
  })

  const transactionHash: Promise<string> = new Promise((resolve, reject) => {
    resolvers.transactionHash = resolve
    rejectors.transactionHash = reject
  })

  const confirmation: Promise<boolean> = new Promise((resolve, reject) => {
    resolvers.confirmation = resolve
    rejectors.confirmation = reject
  })

  const rejectAll = (error: Error) => {
    values(rejectors).map((reject) => {
      // @ts-ignore
      reject(error)
    })
  }

  try {
    logger(Started)
    const txParams: Tx = {
      from: account,
      feeCurrency: feeCurrencyAddress,
      // Hack to prevent web3 from adding the suggested gold gas price, allowing geth to add
      // the suggested price in the selected feeCurrency.
      gasPrice: gasPrice ? gasPrice : '0',
      nonce,
    }

    if (estimatedGas === undefined) {
      estimatedGas = (await estimateGas(tx, txParams)).toNumber()
      logger(EstimatedGas(estimatedGas))
    }

    tx.send({ ...txParams, gas: estimatedGas })
      // @ts-ignore
      .on('receipt', (r: TransactionReceipt) => {
        logger(ReceiptReceived(r))
        if (resolvers.receipt) {
          resolvers.receipt(r)
        }
      })
      .on('transactionHash', (txHash: string) => {
        logger(TransactionHashReceived(txHash))

        if (resolvers.transactionHash) {
          resolvers.transactionHash(txHash)
        }
      })
      .on('confirmation', (confirmationNumber: number) => {
        if (confirmationNumber > 1) {
          // "confirmation" event is called for 24 blocks.
          // if check to avoid polluting the logs and trying to remove the standby notification more than once
          return
        }
        logger(Confirmed)

        if (resolvers.confirmation) {
          resolvers.confirmation(true)
        }
      })
      .on('error', (error: Error) => {
        logger(Failed(error))
        rejectAll(error)
      })
  } catch (error) {
    logger(Exception(error))
    rejectAll(error)
  }

  return {
    receipt,
    transactionHash,
    confirmation,
  }
}
