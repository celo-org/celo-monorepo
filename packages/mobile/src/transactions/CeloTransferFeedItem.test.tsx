import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import { TokenTransactionType } from 'src/apollo/types'
import CeloTransferFeedItem from 'src/transactions/CeloTransferFeedItem'
import { TransactionStatus } from 'src/transactions/types'
import { createMockStore, getMockI18nProps } from 'test/utils'

const SAMPLE_ADDRESS = '0xcc642068bdbbdeb91f348213492d2a80ab1ed23c'

const localAmount = {
  value: '1.23',
  exchangeRate: '0.555',
  currencyCode: 'EUR',
}

describe('CeloTransferFeedItem', () => {
  let dateNowSpy: any
  beforeAll(() => {
    // Lock Time
    dateNowSpy = jest.spyOn(Date, 'now').mockImplementation(() => 1487076708000)
    // set the offset to ALWAYS be Pacific for these tests regardless of where they are run
    // see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTimezoneOffset
    jest.spyOn(Date.prototype, 'getTimezoneOffset').mockImplementation(() => 420)
  })

  afterAll(() => {
    // Unlock Time
    dateNowSpy.mockRestore()
  })

  it('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={createMockStore({})}>
        <CeloTransferFeedItem
          status={TransactionStatus.Complete}
          __typename="TokenTransfer"
          type={TokenTransactionType.Exchange}
          hash={'0x'}
          amount={{ value: '-1.005', currencyCode: 'cGLD', localAmount }}
          address={SAMPLE_ADDRESS}
          comment={''}
          timestamp={1}
          {...getMockI18nProps()}
        />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
