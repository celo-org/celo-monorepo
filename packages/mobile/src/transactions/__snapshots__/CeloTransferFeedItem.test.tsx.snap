// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CeloTransferFeedItem renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  nativeBackgroundAndroid={
    Object {
      "attribute": "selectableItemBackground",
      "type": "ThemeAttrAndroid",
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
        "padding": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingBottom": 2,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          feedItemGoldWithdrawal
        </Text>
      </View>
      <View>
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
              },
            ]
          }
        >
          -
          
          1.005
        </Text>
      </View>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#9CA4A9",
            "fontFamily": "Inter-Regular",
            "fontSize": 14,
            "lineHeight": 18,
          }
        }
      >
        Dec 31 global:at 5:00 PM
      </Text>
    </View>
  </View>
</View>
`;
